### PR TITLE
Feat/her

### PR DESCRIPTION
**Pull request by justindujardin**
I noticed your issue on the main pfrl repo, and tried out your code. I think it was mostly working to begin with, but the reported metric was not great so it was hard to tell. 

I added a mod to let the trainer invoke episode-based callbacks, and then I report the average win % from there. For all bits up to 32 I verified the HER can make progress or get to a 1.0 success rate on my single CPU system <1 to a few minutes depending on the number of bits.

I think that the change to Adam optimizer and DoubleDQN are not needed. 

The decay on exploration noise was really critical. I think for the bitflip envs you only need a few thousand steps to get to > 90% so decaying to 0 over like 5000 steps I found to work really well. Without it, the model spent a long time trying to get passed ~95% success.

Thanks for putting this together! 👏

### CODE REVIEW

**Comment by justindujardin** [comment]
On second thought, maybe the exploration noise isn't a problem, and I messed up by reporting the win rate of the train network. I'm not familiar enough with pfrl to know how to pull those stats out during evaluations.

**Comment by justindujardin** [comment]
I added an eval check, and sure enough the success rate climbs to 100% even with the exploration noise. I'm going to close this PR because despite some better type hints and changes, I don't think anything in your existing PR is broken (other than reporting)

**Comment by prabhatnagarajan** [comment]
Hey, sorry! For some reason I didn't see this pull request until today! Since I borrowed some of your hyperparameters, we can reopen this PR so you get credit when this is merged into PFRL. I guess the changes to `train_agent` aren't needed given the other PR.

**Comment by prabhatnagarajan** [comment]
I'll try and resolve the merge conflicts.

**Comment by prabhatnagarajan** [comment]
It seems like I've ported over most of these changes already and reformatted the files (to pass PFRL's lintchecker), so you'd have to make some changes if we want to merge this PR. Either way, I'll acknowledge you in the pull request and the next PFRL release.

**Comment by justindujardin** [comment]
I'm glad to see the PR is almost complete! If you want to accept my suggested change in your PR, I think it will count as a contribution in the main repo
